language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -346,96 +364,93 @@ impl AbstractState {
+        let max_local = self.max_local;
+        let mut locals = BTreeMap::new();
+        let mut self_removed = BTreeSet::new();
+        let mut other_removed = BTreeSet::new();
+        for idx in self.iter_locals() {
+            let self_value = self.locals.get(&idx);
+            let other_value = other.locals.get(&idx);
+            match (self_value, other_value) {
+                // join error, a resource is available along one path but not the other
+                (None, Some(v)) | (Some(v), None) if v.value.is_possibly_resource() => return None,
+
+                // A reference is assigned on one side, but not the other. Release
+                (Some(v), None) => {
+                    if let AbstractValue::Reference(id) = &v.value {
+                        self_removed.insert(*id);
+                    }
                 }
-                AbstractValue::Value(kind) => {
-                    values.insert(idx.clone(), kind);
+                (None, Some(v)) => {
+                    if let AbstractValue::Reference(id) = &v.value {
+                        other_removed.insert(*id);
+                    }
                 }
+
+                // The local has a value on each side, add it to the state
+                (Some(v1), Some(v2)) => {
+                    checked_verify!(v1 == v2);
+                    checked_verify!(!locals.contains_key(&idx));
+                    locals.insert(idx, v1.clone());
+                }
+
+                (None, None) => (),
             }
         }
+
+        let borrow_graph = {
+            let mut self_graph = self.borrow_graph.clone();
+            let mut other_graph = other.borrow_graph.clone();
+            for reference in self_removed {
+                self_graph.remove_nonce(reference);
+            }
+            for reference in other_removed {
+                other_graph.remove_nonce(reference);
+            }
+            self_graph.join(&other_graph);
+            self_graph
+        };
+        let next_id = self.next_id;
+        let frame_root = self.frame_root;
+
+        Some(Self {
+            locals,
+            max_local,
+            borrow_graph,
+            next_id,
+            frame_root,
+        })
     }
 }
 
 impl AbstractDomain for AbstractState {
     /// attempts to join state to self and returns the result
     fn join(&mut self, state: &AbstractState) -> JoinResult {
-        checked_precondition!(self.is_canonical() && state.is_canonical());
-        // A join failure occurs in each of the following situations:
-        // - a reference or resource is available along one path but not the other
-        // - a value is borrowed along one path but unavailable along the other
-        if self
+        let joined = match Self::join_(self, state) {
+            None => return JoinResult::Error,
+            Some(joined) => joined,
+        };
+        let locals_unchanged = self
```

Comment:
This is a bit fragile because it relies on the fact that `join_` only keeps a key when it is bound to the same value on both sides. Seems worth it to mention this in a comment in case `join_` works differently in the future...

Reply:
> This is a bit fragile because it relies on the fact that join_ only keeps a key when it is bound to the same value on both sides. Seems worth it to mention this in a comment in case join_ works differently in the future...

I'll just update it so I use `iter_locals`. We are more concerned about asymptotic performance, and easy to understand correctness at this point

Reply:
That approach seems very sensible to me.